src/platforms/hosted/jlink.c

```diff
@@ -295,14 +298,21 @@ static bool jlink_get_version(void)
 	if (version_length > sizeof(jlink.fw_version))
 		return false;
 
-	/* Read vesion string directly into jlink.version */
+	/* Read version string directly into jlink.version */
 	bmda_usb_transfer(bmda_probe_info.usb_link, NULL, 0, jlink.fw_version, version_length, JLINK_USB_TIMEOUT);
-	jlink.fw_version[version_length - 1U] = '\0'; /* Ensure null termination */
+	/* Ensure null termination */
+	char *const null_termination = jlink.fw_version + version_length - 1U;
+	*null_termination = '\0';
+
+	/* Replace NULL separating version and copyright string, if it exists */
+	char *const null_separator = strchr(jlink.fw_version, '\0');
+	if (null_separator != NULL && null_separator < null_termination)
```

Comment:
/technically/ not well defined to use `<` on pointers, please prefer `!=` as in:
```suggestion
	if (null_separator != NULL && null_separator != null_termination)
```

Reply:
These are two pointers into the same `jlink.fw_version[]` which is a contiguous char array, and comparing them like this worked (on glibc anyway). We are not asking to compare pointers to different objects. But *technically* replacing this with a not-equal check yields the same behaviour for this exact firmware type and fw_version binary content.

Reply:
It's not about what appears to work, it's about what is actually defined by the C standard and what isn't. Pointer comparison is only completely safe when done as `==` and `!=`. It is *implementation defined* what happens otherwise for `<`, and that means compiler-dependant. If we were comparing `>` then it's covered by the defined semantics, although best practice says to avoid doing this too as it's bugprone.

See the [cppreference page on C comparison operators for more](https://en.cppreference.com/w/c/language/operator_comparison)

Reply:
Sorry, I don't follow. From the same cppreference page:
> if two pointers point to different elements of the same array, the one pointing at the element with the larger index compares greater.

I do NOT read it as "comparing for less-than is UB, always use greater-than and bend over backwards by swapping arguments". Also see C11 6.5.8 Relational operators from $60 pdf or http://port70.net/%7Ensz/c/c11/n1570.html#6.5.8
>  If the objects pointed to are members of the same aggregate object, pointers to structure members declared later compare greater than pointers to members declared earlier in the structure, and pointers to array elements with larger subscript values compare greater than pointers to elements of the same array with lower subscript values.

And especially

> Each of the operators < (less than), > (greater than), <= (less than or equal to), and >= (greater than or equal to) shall yield 1 if the specified relation is true and 0 if it is false. The result has type int.

I spent an hour writing a test snippet so people can check more easily. Generated assembly is not identical, but runtime behaviour is. https://godbolt.org/z/n5WEon9sM
BMDA is supported AFAIK on MSYS2, Linux, Darwin, but isn't on 16-bit segmented memory arches.

Looking at URB responses in Wireshark I see a 112-byte response with ~16 NULL bytes at the tail of version string, probably because their firmware simply returns then entire chunk from .rodata (or configuration space) of fixed length. I don't want to chase these NULL bytes, just the one separating version from copyright (if it always exists).

Reply:
What you're observing is what we mentioned about "implementation defined what happens". Note that the text on pointers never talks about `<`, only `>` - you are inferring that defining greater than defines less than, but this is not in the plain text of the standard and while that is how many implementations do define it, they don't have to.

Please just use `!=` - it is fully defined for pointers for both C and C++ and will be safer into the future of the code base. This is what the recommended best practice is, this is the relation that is defined on std::span<> and C++ iterator comparisons too. `<` is not as is `>` because they are unable to be fully well defined as a relation on pointers. There is good reason for this. We are trying in this case to avoid someone in the future editing this code and invalidating the precondition that the pointers have be in the same buffer so we can avoid all the pitfalls of the inequality comparison operators.

`strchr()` won't go past the first `\0` so that's not a problem - for all J-Link firmware that's working the way you've observed (that 112-byte response), `null_separator != null_termination` will always be true - but more importantly, will also always be well defined.

Reply:
Thank you for the detailed explanation. Will prefer pointer equality checks where possible over less-than/greater-than. I wonder if compilers have a diagnostic for that.